gnucash/gnome-utils/gnc-gtk-utils.c

```diff
@@ -332,3 +332,320 @@ gnc_get_dialog_widget_from_id (GtkDialog *dialog, const gchar *id)
     GtkWidget *content_area = gtk_dialog_get_content_area (dialog);
     return find_widget_func (content_area, id);
 }
+
+
+/** Disable all the actions in a simple action group
+ *
+ *  @param action_group The GSimpleActionGroup
+ */
+void
+gnc_disable_all_actions_in_group (GSimpleActionGroup *action_group)
+{
+    gchar **actions;
+    gint num_actions;
+
+    g_return_if_fail (action_group != NULL);
+
+    actions = g_action_group_list_actions (G_ACTION_GROUP(action_group));
+    num_actions = g_strv_length (actions);
+
+    // Disable the actions
+    for (gint i = 0; i < num_actions; i++)
+    {
+        GAction *action = g_action_map_lookup_action (G_ACTION_MAP(action_group),
+                                                      actions[i]);
+        g_simple_action_set_enabled (G_SIMPLE_ACTION(action), FALSE);
+    }
+    g_strfreev (actions);
+}
+
+
+static void
+add_accel_for_menu_lookup (GtkWidget *widget, gpointer user_data)
+{
+    if (GTK_IS_MENU_ITEM(widget))
+    {
+        GtkMenuItem* menuItem = GTK_MENU_ITEM(widget);
+        GtkWidget* subMenu = gtk_menu_item_get_submenu (menuItem);
+        GtkWidget *accel_label = gtk_bin_get_child (GTK_BIN(widget));
+
+        if (accel_label)
+        {
+            guint key;
+            GdkModifierType mods;
+
+            gtk_accel_label_get_accel (GTK_ACCEL_LABEL(accel_label), &key, &mods);
+
+            if (key > 0)
+                gtk_widget_add_accelerator (GTK_WIDGET(widget), "activate",
+                                            GTK_ACCEL_GROUP(user_data),
+                                            key, mods, GTK_ACCEL_VISIBLE);
+        }
+        if (GTK_IS_CONTAINER(subMenu))
+            gtk_container_foreach (GTK_CONTAINER(subMenu),
+                                   add_accel_for_menu_lookup, user_data);
+    }
+}
+
+/** Add accelerator keys for menu item widgets
+ *
+ *  @param menu The menu widget.
+ *
+ *  @param accel_group The accelerator group to use.
+ */
+void
+gnc_add_accelerator_keys_for_menu (GtkWidget *menu, GtkAccelGroup *accel_group)
+{
+    g_return_if_fail (GTK_IS_WIDGET(menu));
+    g_return_if_fail (accel_group != NULL);
+
+    gtk_container_foreach (GTK_CONTAINER(menu), add_accel_for_menu_lookup, accel_group);
+}
+
+
+static gpointer
+find_menu_item_func (GtkWidget *widget, const gchar *action_name)
+{
+    GtkWidget *ret = NULL;
+
+    if (GTK_IS_MENU_ITEM(widget))
+    {
+        const gchar *a_name = g_object_get_data (G_OBJECT(widget), "myaction-name");
```

Comment:
Every GtkMenuItem also implements the GtkActionable interface. So I think it should be possible to simply do
```
const gchar *a_name = gtk_actionable_get_action_name (widget);
```

This may need a typecast for widget to GTK_ACTIONABLE.

However as to me it looks like the action name is readily available, I wonder if you really need to set the myaction-name property on each menu item separately.

Reply:
That's what I thought but unfortunately when the menu is built from the model the GTK_ACTIONABLE is not populated. So to be able to find the menu items to hide them as required I needed another way to find them other that the label that is why I added this. Using inspector, all the menu items that are added by the model have the 'action-name' field blank unlike the ones I have built for the Aqbanking menu items that have the full action name.
I still maybe missing some thing...

Reply:
That's surprising.
Perhaps this is because we are still loading old-style ui files instead of true glade files ?
If I load one of our ui files in glade only the toolbar appears, not the menubar. Note for example in gnc-main-window.ui the toolbar is defined as an `object`, and the actions are referred to via `action-name`. In the menubar definitions these are referred to via `action`. Perhaps that's the issue ?

Reply:
If so, this means we may have to redo the ui files in glade or convert them to glade format in some way. Might as well change the extension to .glade then. Some may even be merged with glade files for the same object. For example there's both a gnc-plugin-page-budget.glade and a gnc-plugin-page-budget.ui. If we convert the latter to glade format, it could be added in the former.

Reply:
More likely it's because we're still using GtkMenu instead of GMenu. That's not a simple change: It also requires GApplication and removing GtkOsxApplication and maybe modifying some of our other macOS customizations.

Reply:
Perhaps. A simple test should give more details.
@Bob-IT can you experiment with setting property `action-name` instead (or in addition to) the `action` property ? If that fails to parse you may have to try and recreate one menubar in glade, such that it's described as an objects hierarchy (similar to how the toolbar is already described), and then check in the inspector whether the action-name property is set ?

Reply:
That's the workaround I was going to suggest if we couldn't make the glade ui files work :wink: 

More on this later.